### ISSUE TITLE
[BUG REPORT] References are rewritten on commit

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#53]
**Describe the bug**
With the new weakref logic, ideally, hangar's user-facing objects such as dataset/metadata are weak references to the actual object and it gets removed on `checkout.close()`. But since we are rewriting `self._datasets` & `self._metadata` here
https://github.com/tensorwerk/hangar-py/blob/deeeafc96e683317578a81233fadb0248a2c39c4/src/hangar/checkout.py#L336-L341
on `commit` (not only on `close` since `self.__setup()` gets called on `commit` as well)
https://github.com/tensorwerk/hangar-py/blob/deeeafc96e683317578a81233fadb0248a2c39c4/src/hangar/checkout.py#L387
the references are gone after the commit. In a nutshell, the issue is: User can't make changes after the commit even if the checkout is not closed yet.

**Severity**

Select an option:
- [ ] Data Corruption / Loss of Any Kind
- [x] Unexpected Behavior, Exceptions or Error Thrown
- [ ] Performance Bottleneck

**To Reproduce**
Steps to reproduce the behavior, minimal example code preffered:
```python
co = repo.checkout(write=True)
dset = co.datasets.init_dataset('dset', prototype=array5by7)
dset.add(array5by7, '1')
co.commit()
dset.add(array5by7, '2')  # this raises ReferenceError since the reference to dset is gone
co.close()
```

**Expected behavior**
User should be able to add data after the commit. References should be released only after the close. Perhaps we should split `__setup()` function into two with very specific functionalities.

**Comment by rlizzo** [comment]
Hey @hhsecond, thanks for spotting this! I've got a fix up in #54 if you want to take a look?

**Comment by rlizzo** [comment]
Fixed in #54
